lib/spack/spack/cmd/config.py

```diff
@@ -135,11 +164,150 @@ def config_list(args):
     print(' '.join(list(spack.config.section_schemas)))
 
 
+def config_add(args):
+    """Add the given configuration to the specified config scope
+
+    This is a stateful operation that edits the config files under the hood"""
+    if not (args.file or args.value):
+        tty.error("No changes requested. Specify a file or value.")
+        setup_parser.add_parser.print_help()
+        exit(1)
+
+    scope, section = _get_scope_and_section(args)
+
+    # Updates from file
+    if args.file:
+        # Get file as config dict
+        with open(args.file, 'r') as f:
+            data = syaml.load_config(f)
+
+        if 'env' in data or 'spack' in data:
+            # Validate as an environment
+            # use config_dict to remove top-level key
+            spack.environment.validate(data, 'spack.yaml')
+            config_dict = spack.environment.config_dict(data)
+        else:
+            # Separate config file
+            # Use schema according to first key (they each only accept one key)
+            # Do not remove top key
+            spack.config.validate(
+                data, spack.config.section_schemas[next(iter(data))])
```

Comment:
This will fail with an unintuitive error message if the data is an empty list/dict, or if the key is not in `section_schemas`.  Look at `_validate_section_name` in `spack.config`.  I think the best way to handle this and the prior comment is to add another method to return a schema name given some YAML (and raise `ConfigError` using `_validate_schema_name` if it's not valid)